httpx/client.py

```diff
@@ -172,42 +156,62 @@ def merge_headers(
 
         return response
 
-    def _resolve_dispatcher(
+    async def send(
         self,
         request: AsyncRequest,
+        *,
+        stream: bool = False,
         auth: AuthTypes = None,
-        trust_env: bool = False,
         allow_redirects: bool = True,
-    ) -> AsyncDispatcher:
-        dispatcher: AsyncDispatcher = RedirectDispatcher(
-            next_dispatcher=self.dispatch,
-            base_cookies=self.cookies,
-            allow_redirects=allow_redirects,
+        verify: VerifyTypes = None,
+        cert: CertTypes = None,
+        timeout: TimeoutTypes = None,
+        trust_env: bool = None,
+    ) -> AsyncResponse:
+        if request.url.scheme not in ("http", "https"):
+            raise InvalidURL('URL scheme must be "http" or "https".')
+
+        get_response = functools.partial(
+            self.get_response, stream=stream, verify=verify, cert=cert, timeout=timeout
         )
 
-        username: typing.Optional[typing.Union[str, bytes]] = None
-        password: typing.Optional[typing.Union[str, bytes]] = None
-        if auth is None:
-            if request.url.username or request.url.password:
-                username, password = request.url.username, request.url.password
-            elif trust_env:
-                netrc_login = get_netrc_login(request.url.authority)
-                if netrc_login:
-                    username, _, password = netrc_login
-        else:
-            if isinstance(auth, tuple):
-                username, password = auth[0], auth[1]
-            elif callable(auth):
-                dispatcher = CustomAuthDispatcher(
-                    next_dispatcher=dispatcher, auth_callable=auth
-                )
-
-        if username is not None and password is not None:
-            dispatcher = BasicAuthDispatcher(
-                next_dispatcher=dispatcher, username=username, password=password
+        get_response = functools.partial(
+            redirect(allow_redirects, cookies=self.cookies), get_response=get_response
```

Comment:
Personally I'm not sure if this is clearer that the class based approach, I go into more detail on the definition for `Redirect` below, but here it seems we're "forcing" redirect to be a callable for no particular gain?

Keen on hearing other peoples thoughts though 🙂

Reply:
The latest iteration provides a `BaseMiddleware` class, which requires subclasses to explicitly implement `__call__()`. Would you say this make the callable-based implementation clearer?